EG-1856: Allow users to specify Dockerform ports

palade opened the pull request:
Dockerform allocates predefined port numbers instead of using the ones specified in the `node.conf`. Users should be allowed to specify those ports

chrisr3 commented on cordformation/src/main/kotlin/net/corda/plugins/RpcSettings.kt:
```diff
@@ -34,6 +39,9 @@ class RpcSettings {
     fun adminAddress(value: String) {
+        try {
+            adminPort = URI(null, value, null, null, null).port
+        } catch (ex: URISyntaxException) { }
```
Same as above. And maybe make a function that both properties can use for this?

palade commented on cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt:
```diff
@@ -99,7 +99,7 @@ open class Dockerform @Inject constructor(objects: ObjectFactory) : Baseform(obj
                             "$nodeBuildDir/drivers:/opt/corda/drivers"
                     ),
-                    "ports" to listOf(it.rpcPort, it.config.getInt("sshd.port")),
+                    "ports" to listOf(it.rpcPort(), it.config.getInt("sshd.port")),
```
> Does `sshd.port` need to be different for each node too?

Not necessarily, but a user may want to see the port number that he set in the `build.conf`

palade commented:
> Looks OK. Is there any way that your new test can also check whether the RPC ports have been set correctly? After all, that _is_ the point of this PR.

Added a unit test to check for correct parsing of the RPC  port. Moved the parsing function in separate file and added several unit tests for that function as well

chrisr3 commented on cordformation/src/test/kotlin/net/corda/plugins/DockerformTest.kt:
```diff
@@ -111,5 +114,23 @@ class DockerformTest : BaseformTest() {
+        val notaryConfig = ConfigFactory.parseFile(notaryConfigPath.toFile())
+        assertTrue(notaryConfig.hasPath("rpcSettings.address"))
+        assertTrue(notaryConfig.hasPath("rpcSettings.adminAddress"))
+        assertEquals(ConfigurationUtils.parsePort(notaryConfig.getString("rpcSettings.address")), 10003)
```
Funny thing - this `assertEquals()` function is really:
```kotlin
assertEquals(expectedValue, actualValue)
```
i.e. the assertion should be written as:
```kotlin
assertEquals(10003, ConfigurationUtils.parsePort(notaryConfig.getString("rpcSettings.address")))
```
(FYI different frameworks put these parameters in different orders. I don't know why....)

palade replied:
Fixed now :-)

chrisr3 commented on cordformation/src/test/kotlin/net/corda/plugins/ConfigurationUtilsTest.kt:
```diff
@@ -0,0 +1,23 @@
+
+    @Test
+    fun `check correct port value parsing`() {
+        assertEquals(ConfigurationUtils.parsePort("localhost:10000"), 10000)
```
Oops, `assertEquals(expectedValue, actualValue)`...

chrisr3 commented on cordformation/src/main/kotlin/net/corda/plugins/ConfigurationUtils.kt:
```diff
@@ -0,0 +1,19 @@
+            return try {
+                URI(null, address, null, null, null).port
+            } catch (ex: URISyntaxException) {
+                throw InvalidUserDataException("Invalid host and port syntax for RPC address, expected host:port. Using default value")
```
This exception will kill Gradle, which is fine. It's just that the message shouldn't therefore say that the default value will be used.
